w32_common: refactor and improve window state handling

pavelxdd opened the pull request:
Refactored `reinit_window_state` function:
Moved all the code related to fitting window on screen in a separate function `fit_window_on_screen`.
Also simplified that code a lot to make it easier to read and added a few debug messages.

Fixed #4753 by disabling the above mensioned code in `fit_window_on_screen` for a maximized window, since maximized window should already fit on the screen just fine.
It should be noted that this bug was only reproducible with `--fit-border` option which is enabled by default. The cause of the bug is that after calling the `add_window_borders` for a maximized window, the rect in result is slightly larger than the screen rect, which is okay, `SetWindowPos` will interpret it as a maximized state later, so no auto-fitting to screen size is needed here.

In `reinit_window_state` debug message `reset window bounds` now reports client area size and position, instead of window area size and position. This is done for consistency with debug messages in handling
fullscreen state above in this function, since they also print window bounds of the client area.

All in all, the `reinit_window_state` is much shorter now.

shinchiro commented:
Tested this patch. After returned from first fullscreen, if I try to enter fullscreen again, this happened.
![image](https://user-images.githubusercontent.com/9623615/33801876-45a85efc-dda4-11e7-8256-a2d83c6e3b60.png)

pavelxdd commented:
@shinchiro, thanks, fixed.

`SetWindowLongPtrW` sometimes triggers `WM_SIZE` events, which updates the `w32->dw` and `w32->dh` values, so they should be saved beforehand.

pavelxdd commented:
Oops, forgot to move the changed size and pos from rect back to `w32` struct. Updated again.

rossy commented on video/out/w32_common.c:
```diff
@@ -784,8 +785,10 @@ static void fit_window_on_screen(struct vo_w32_state *w32, RECT *rc)
+    // Use virtual screen size, since the mpv window can be resized
+    // to cover mupltiple monitors
+    const int screen_w = GetSystemMetrics(SM_CXVIRTUALSCREEN);
+    const int screen_h = GetSystemMetrics(SM_CYVIRTUALSCREEN);
```
I don't think mpv should use the virtual screen size here, since this will affect the window size with newly opened files as well as when mpv leaves fullscreen. Maybe mpv should just skip the fit-to-screen logic when leaving fullscreen, but use it in all other cases.

pavelxdd replied:
The old behavior also affected newly opened files as well as leaving the fullscreen state, I'm not sure how using virtual size making things worse than now, can you elaborate please? I agree that skipping that function when leaving fullscreen might be a good idea. Should I leave the virtual screen size here anyway?

rossy commented:
Looks good. ``fit_window_on_screen`` is definitely clearer than the old logic. I'm not sure using the virtual screen size is the best solution to #5215 though, since it also affects the window size when loading new files.

pavelxdd commented:
Just for the reference, this function doesn't affect `autofit` options in any way, so it shouldn't have any side effects.

pavelxdd commented:
Using virtual screen size is the only way to fix #5215 , I believe, because `reinit_window_state` is called not only when switching fullscreen state, but on other events as well, like toggling border or ontop options. So skipping `fit_window_on_screen` on leaving fullscreen is not an option, and doing this is going to bring inconsistency to mpv window behavior.

rossy commented:
There is definitely one side-effect. If you don't use --autofit and you open a file that's larger than the current monitor, mpv will size the window to the virtual screen with these patches. Without the patch, it sizes the window to the current monitor.

I do think #5215 is a valid bug, but maybe it's better solved by skipping the ``fit_window_on_screen`` logic in some cases. It definitely seems odd that ``fit_window_on_screen`` is applied when toggling ontop.

pavelxdd commented:
> Without the patch, it sizes the window to the current monitor.

The question is, does the user on a multi-monitor system really expect this behavior? What if someone has 9 screens as a one huge monitor (3x3) and doesn't want to use `--fs-screen=all` for some reason?

At least this should be optional I think, like I've already suggested in the issue #5215 
I don't like the idea of skipping this function, because in this case Windows will handle unusual window sizes anyway, and it might contradict with `--fit-border` option.

rossy commented:
> The question is, does the user on a multi-monitor system really expect this behavior? 

I think so. I think the common case is to have a pair of monitors, a primary monitor and auxiliary monitor or some other device like a TV or a projector. You don't really want programs to decide to span multiple-monitors of their own accord with these setups, since it's annoying to have to size them back down to the monitor you are using, especially if it sizes the window back up when opening new files, and especially if the monitors aren't actually next to each other, like with a projector setup.

It seems like this is trading one strange behaviour (mpv restoring to a different size when leaving fullscreen) with another (mpv spanning multiple monitors on startup.) But the simple way to make mpv restore to the same size after leaving fullscreen is to skip the fit-to-screen logic. It's odd for something like toggling fullscreen (or toggling ontop) to change the user-specified window size. Toggling fullscreen should work like toggling maximize. (We could use Get/SetWindowPlacement for this.)

rossy commented on video/out/w32_common.c:
```diff
@@ -766,35 +766,80 @@ static DWORD update_style(struct vo_w32_state *w32, DWORD style)
-    bool toggle_fs = w32->current_fs != new_fs;
-    w32->current_fs = new_fs;
-    w32->toggle_fs = false;
+    MP_VERBOSE(w32, "adjusting window size to fit it on the screen\n");
```
This is a bit too chatty. You can tell if the window size was adjusted from the "adjusted window size" message at the end of the function. Verbose-mode is very noisy as-is.

pavelxdd commented:
There's another issue (#2202) which could benefit from using `Get/SetWindowPlacement`, I planned to address this problem a bit later, to not make a mess out of this PR.

I agree with you, that this is expected behavior for a user with dual-screen setup. I'll see what can be done with this when I get home.

pavelxdd commented:
Updated. Please read the commit messages.

I'm using `RECT` now for storing screen and window size & pos, since it's much easier to use them this way in WinAPI functions.
Split the `fit_window_on_screen` into two functions: `fit_rect` is a helper function to fit one rect on another.

Splitting the `reinit_window_state` into 4 functions allowed to skip the certain parts on updating the `ontop` and `border` properties.

I'm skipping the `fit_window_on_screen` when toggling fullscreen, so the issue we discussed above is fixed now without using the virtual screen size.

pavelxdd commented:
Found another minor issue: since the `fit_window_on_screen` was skipped on leaving fullscreen, when a new video file was opened (via drag-n-drop) while the window was in fullscreen, the `fit_window_on_screen` wasn't called to adjust the new window size.

Fixed it by adding a new flag `fit_on_screen`. When this flag is set, the `fit_window_on_screen` will be called even when leaving fullscreen. I set this flag in `gui_thread_reconfig` on window initialization and when the video size was reset.

Refactored `gui_thread_reconfig` according to my recent window RECT changes.

rossy commented on video/out/w32_common.c:
```diff
@@ -62,6 +62,11 @@ typedef enum MONITOR_DPI_TYPE {
 #endif
 
+#define rect_x(r) ((int)((r).left))
+#define rect_y(r) ((int)((r).top))
```
rect_x/rect_y seem a bit redundant/unnecessary.

rossy commented on video/out/w32_common.c:
```diff
@@ -1563,42 +1556,32 @@ static int gui_thread_control(struct vo_w32_state *w32, int request, void *arg)
-        s[1] = w32->current_fs ? w32->prev_height : w32->dh;
-        return VO_TRUE;
-    }
+    case VOCTRL_GET_UNFS_WINDOW_SIZE:
```
Not sure it's best to collapse the VOCTRL_GET_UNFS_WINDOW_SIZE and VOCTRL_SET_UNFS_WINDOW_SIZE cases. There's only a couple of lines of shared logic, so it's probably clearer to keep them separate.

rossy commented on video/out/w32_common.c:
```diff
@@ -128,6 +128,9 @@ struct vo_w32_state {
     int high_surrogate;
 
+    // Whether to fit the window on screen on next window state updating
+    bool fit_on_screen;
```
Fine for now, but in future it might be better to solve this by applying the fit-on-screen logic to prev_windowrc when it's updated.

pavelxdd replied:
Maybe, but it'll require storing the virtual screen size in a separate RECT, i.g. `vscreenrc`, instead of `screenrc`, for the case when `--fs-screen=all` is set. I still think that the fit on screen logic should stay as it is now, to make sure the window styles are up to date at the moment of adjusting the window bounds.

rossy commented:
Looks fine as far as I can tell, except for a couple of minor issues that I commented on. I like that it's using RECT now, since it simplifies a lot of code.

pavelxdd commented:
Updated.
